CRM/Activity/Tokens.php

```diff
@@ -138,75 +103,122 @@ public function prefetch(TokenValueEvent $e) {
    * @throws \CRM_Core_Exception
    */
   public function evaluateToken(TokenRow $row, $entity, $field, $prefetch = NULL) {
-    // maps token name to api field
-    $mapping = [
-      'activity_id' => 'id',
-    ];
-
     $activityId = $row->context[$this->getEntityContextSchema()];
 
-    $activity = $prefetch['activity'][$activityId];
-
-    if (in_array($field, ['activity_date_time', 'created_date', 'modified_date'])) {
-      $row->tokens($entity, $field, \CRM_Utils_Date::customFormat($activity[$field]));
-    }
-    elseif (isset($mapping[$field]) and (isset($activity[$mapping[$field]]))) {
-      $row->tokens($entity, $field, $activity[$mapping[$field]]);
-    }
-    elseif (in_array($field, ['activity_type'])) {
-      $row->tokens($entity, $field, $this->activityTypes[$activity['activity_type_id']]);
-    }
-    elseif (in_array($field, ['status'])) {
-      $row->tokens($entity, $field, $this->activityStatuses[$activity['status_id']]);
-    }
-    elseif (in_array($field, ['campaign'])) {
-      $row->tokens($entity, $field, $this->campaigns[$activity['campaign_id']]);
+    if (!empty($this->getDeprecatedTokens()[$field])) {
+      $realField = $this->getDeprecatedTokens()[$field];
+      parent::evaluateToken($row, $entity, $realField, $prefetch);
+      $row->format('text/plain')->tokens($entity, $field, $row->tokens['activity'][$realField]);
     }
     elseif (in_array($field, ['case_id'])) {
       // An activity can be linked to multiple cases so case_id is always an array.
       // We just return the first case ID for the token.
-      $row->tokens($entity, $field, is_array($activity['case_id']) ? reset($activity['case_id']) : $activity['case_id']);
+      // this weird hack might exist because apiv3 is weird &
```

Comment:
I haven't reviewed this fully yet just commenting on this - it's not api. You can have a single activity on multiple cases. Link-cases is an example in core. But returning a random case id when there's more than one maybe isn't a useful token.

Reply:
@demeritcowboy do you think I can just remove it? It's pre-existing but I'd be happy to kill it.... although maybe I need to socialise that / make it a follow up?

Reply:
When there's only one case it lives in, which is most of the time, it could be a useful token.

I am curious about hearing more about the use-cases for activity print/merge in general.

Reply:
@demeritcowboy yeah that stuff was added by @aydun 

I'd like to see if we can resolve this to get the existing tokens working & move that question to gitlab (since I just sorts keep it working here)